src/Processors/LimitTransform.cpp

```diff
@@ -130,7 +225,234 @@ LimitTransform::Status LimitTransform::prepare()
+    queue.emplace_back(std::move(element));
+}
+
+/// When limit and offset are negative, pop chunks from the queue if all input ports are finished.
+/// Cut operation will not occur more than twice.
+LimitTransform::QueueElement LimitTransform::queuePop()
+{
+    if (limit_is_unreachable && canPopWithoutCut())
+    {
+        QueueElement pop(std::move(queue.front()));
+        queue.pop_front();
+        rows_in_queue -= pop.chunk.getNumRows();
+        return pop;
+    }
+
+    QueueElement pop(std::move(queue.front()));
+    queue.pop_front();
+    rows_in_queue -= pop.chunk.getNumRows();
+
+    if (rows_in_queue >= offset)
+    {
+        if (rows_in_queue + pop.chunk.getNumRows() <= rows_to_keep)
+            ///              <--------> rows_in_queue
+            /// <---------------------> rows_to_keep
+            /// ^limit            ^offset
+            ///      <-------> pop
+            ///      <-------> to return
+            return pop;
+
+        /// Need to cut chunk
+        UInt64 num_columns = pop.chunk.getNumColumns();
+        UInt64 num_rows = pop.chunk.getNumRows();
+        Columns columns = pop.chunk.detachColumns();
+
+        ///                  <--------> rows_in_queue
+        ///            <--------------> rows_to_keep
+        ///            ^limit      ^offset
+        /// <----------------> pop
+        ///            <-----> to return
+        UInt64 diff = num_rows + rows_in_queue - rows_to_keep;
+        for (UInt64 i = 0; i < num_columns; ++i)
+            columns[i] = columns[i]->cut(diff, num_rows - diff);
+        pop.chunk.setColumns(std::move(columns), num_rows - diff);
+        return pop;
+    }
+
+    /// Need to cut chunk
+    UInt64 num_columns = pop.chunk.getNumColumns();
+    UInt64 num_rows = pop.chunk.getNumRows();
+    Columns columns = pop.chunk.detachColumns();
+    UInt64 start;
+    UInt64 length;
+
+    if (rows_in_queue + num_rows > rows_to_keep)
+    {
+        ///                   <------------------> rows_in_queue
+        /// <----------------> pop
+        ///    ^limit    ^offset
+        ///     <--------> to return (limit is reachable)
+        /// <------------> to return (limit is unreachable)
+        /// When limit is unreachable, need to return all rows before offset
+        if (limit_is_unreachable)
+            start = 0;
+        else
+            start = rows_in_queue + num_rows - rows_to_keep;
+    }
+    else
+    {
+        ///                      <------------------> rows_in_queue
+        ///     <----------------> pop
+        /// ^limit    ^offset
+        ///     <-----> to return
+        start = 0;
+    }
+
+    length = rows_in_queue + num_rows - offset - start;
+    for (UInt64 i = 0; i < num_columns; ++i)
+        columns[i] = columns[i]->cut(start, length);
+    pop.chunk.setColumns(std::move(columns), length);
+    return pop;
+}
+
+/// Discard closed output ports' chunks in the queue.
+void LimitTransform::skipFinishedPorts()
```

Comment:
```suggestion
void LimitTransform::skipChunksForFinishedOutputPorts()
```

Then probably a reader will have no need go to the member definition to understand what it's doing